### PR TITLE
feature: Adding serial inference pipeline support to RegisterModel Step

### PR DESCRIPTION
**Pull request by sreedes**
*Issue #, if available:*
#2291, #2014, #2485

*Description of changes:*
This change enables multiple models to be packed together and registered as a single model package version, so that the models could be run as an inference pipeline.
- Updated the inputs for RegisterModelStep to also take a model object or a PipelineModel for Serial Inference pipeline
-  Performs repack in a loop for containers in a serial inference pipeline, if required.
- Ability to invoke RegisterModelStep without an Estimator.
- The environment variables are auto-updated by the framework classes into the model package

*Testing done:*
test_register_model_sip - Able to add more than 1 model during registration of model package.
## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [X] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [X] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [X] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [X] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/workflow/_utils.py [hunk 1]
```diff
@@ -205,18 +205,19 @@ class _RegisterModelStep(Step):
...
     def __init__(
         self,
         name: str,
         estimator: EstimatorBase,
-        model_data,
         content_types,
         response_types,
         inference_instances,
         transform_instances,
+        model_data=None,
```
Comment (danabens) [inline src/sagemaker/workflow/_utils.py:220]: is this change necessary?

File: src/sagemaker/workflow/_utils.py [hunk 2]
```diff
@@ -225,6 +226,7 @@ def __init__(
         compile_model_family=None,
         description=None,
         depends_on: List[str] = None,
+        container_def_list=None,
         **kwargs,
     ):
         """Constructor of a register model step.
```
Comment (danabens) [inline src/sagemaker/workflow/_utils.py:232]: nit - see below that `model_data` is missing a type in the argument description. i.e. `model_data: the S3...` please add `model_data (str): the S3...`

File: src/sagemaker/workflow/_utils.py (hunk 1 above)
Comment (sreedes) [reply src/sagemaker/workflow/_utils.py:221]: > is this change necessary?

Yes, now that we are passing model list, this input may be empty, hence marked it as None.

File: src/sagemaker/workflow/_utils.py (hunk 2 above)
Comment (sreedes) [reply src/sagemaker/workflow/_utils.py:233]: sure i shall update.

**Review by danabens — COMMENTED** [review]
So no changes needed to RepackModelStep?

File: src/sagemaker/workflow/_utils.py [hunk 3]
```diff
@@ -284,41 +295,55 @@ def __init__(
...
-
-        model_package_args = model._get_model_package_args(
+                # create_model wants the estimator to have a model_data attribute...
+                self.estimator._current_job_name = model_name
+
+                # placeholder. replaced with model_data later
+                output_path = self.estimator.output_path
+                self.estimator.output_path = "/tmp"
+
+                # create the model, but custom funky framework stuff going on in some places
```
Comment (jerrypeng7773) [inline src/sagemaker/workflow/_utils.py:316]: shall we remove this comment?

**Review by jerrypeng7773 — DISMISSED** [review]

File: src/sagemaker/workflow/_utils.py [hunk 4]
```diff
@@ -17,9 +17,10 @@
 import shutil
 import tarfile
 import tempfile
-
 from typing import List
 
+import sagemaker
```
Comment (ahsan-z-khan) [inline src/sagemaker/workflow/_utils.py:22]: why are we importing `sagemaker` here?

**Review by ahsan-z-khan — APPROVED** [review]
